File: MB-Whatsapp/confidence/confidence-engine.ts
import { createConfidenceAuditEvent, InMemoryConfidenceAuditSink } from "./confidence-audit";
import { explainConfidence } from "./confidence-explainer";
import { DEFAULT_CLINICAL_CONFIDENCE_FLAGS } from "./confidence-flags";
import { DEFAULT_CONFIDENCE_POLICY } from "./confidence-policy";
import { calculateConfidenceScore } from "./confidence-score";
import { buildEscalationRecommendation } from "./escalation-recommendation";
import { evaluateEvidence } from "./evidence-evaluator";
import { estimateHallucinationRisk } from "./hallucination-risk";
import { detectMultimodalConflict } from "./multimodal-conflict";
import { evaluateProviderConsistency } from "./provider-consistency";
import { evaluateSafeDisplay } from "./safe-display";
import { calculateUncertaintyScore } from "./uncertainty-score";
import type { ClinicalConfidenceFlags, ClinicalConfidenceInput, ClinicalConfidenceResult, ConfidencePolicy } from "./types";

export class ClinicalConfidenceEngine {
  constructor(
    private readonly flags: ClinicalConfidenceFlags = DEFAULT_CLINICAL_CONFIDENCE_FLAGS,
    private readonly policy: ConfidencePolicy = DEFAULT_CONFIDENCE_POLICY,
    private readonly auditSink = new InMemoryConfidenceAuditSink(),
  ) {}

  evaluate(input: ClinicalConfidenceInput): ClinicalConfidenceResult {
    const evidence = evaluateEvidence(input);
    const providerConsistency = this.flags.providerConsistencyEnabled
      ? evaluateProviderConsistency(input.provider_outputs)
      : { providers_compared: 0, consistency_score: 1, conflicts_detected: false, unresolved_conflicts: [] };
    const multimodal = this.flags.multimodalEnabled
      ? detectMultimodalConflict(input)
      : { multimodal_conflict_detected: false, conflicts: [] };
    const hallucinationRisk = this.flags.hallucinationCheckEnabled
      ? estimateHallucinationRisk({
          evidence,
          providerConsistency,
          multimodalConflictDetected: multimodal.multimodal_conflict_detected,
        })
      : {
          risk_level: "low" as const,
          unsupported_claims_detected: false,
          evidence_missing: false,
          multimodal_inconsistency: false,
          provider_divergence: false,
          escalation_required: false,
        };
    const uncertaintyScore = calculateUncertaintyScore({ evidence, providerConsistency, hallucinationRisk });
    const confidenceScore = calculateConfidenceScore({
      evidence,
      providerConsistency,
      uncertaintyScore,
      hallucinationRisk,
    });
    const escalation = buildEscalationRecommendation({
      confidenceScore,
      uncertaintyScore,
      hallucinationRisk,
      evidenceCompleteness: evidence.evidence_completeness,
      providerConflicts: providerConsistency.unresolved_conflicts,
      multimodalConflicts: multimodal.conflicts,
      policy: this.policy,
    });
    const safeDisplay = evaluateSafeDisplay({
      confidenceScore,
      uncertaintyScore,
      hallucinationRisk,
      escalationRecommended: escalation.escalation_recommended,
      flags: this.flags,
      policy: this.policy,
    });
    const baseResult = {
      trace_id: input.trace_id,
      confidence_score: confidenceScore,
      uncertainty_score: uncertaintyScore,
      hallucination_risk: hallucinationRisk,
      evidence_completeness: evidence.evidence_completeness,
      provider_consistency: providerConsistency,
      multimodal_conflict_detected: multimodal.multimodal_conflict_detected,
      escalation_recommended: this.flags.enabled && this.flags.autoEscalationEnabled ? escalation.escalation_recommended : false,
      escalation_reason: escalation.escalation_reason,
      safe_to_display: safeDisplay.safe_to_display,
      display_restrictions: safeDisplay.display_restrictions,
      audit_ref: `${input.trace_id}:clinical-confidence`,
      created_at: new Date().toISOString(),
    };
    const result: ClinicalConfidenceResult = {
      ...baseResult,
      confidence_explanation: explainConfidence({ result: baseResult, evidence }),
    };
    this.auditSink.append(createConfidenceAuditEvent(result));
    return result;
  }

  auditEvents() {
    return this.auditSink.list();
  }
}

File: MB-Chat/confidence_py/confidence_engine.py
from __future__ import annotations

from datetime import UTC, datetime

from .confidence_audit import InMemoryConfidenceAuditSink, create_confidence_audit_event
from .confidence_explainer import explain_confidence
from .confidence_flags import load_clinical_confidence_flags
from .confidence_policy import DEFAULT_CONFIDENCE_POLICY
from .confidence_score import calculate_confidence_score
from .escalation_recommendation import build_escalation_recommendation
from .evidence_evaluator import evaluate_evidence
from .hallucination_risk import estimate_hallucination_risk
from .multimodal_conflict import detect_multimodal_conflict
from .provider_consistency import evaluate_provider_consistency
from .safe_display import evaluate_safe_display
from .types import (
    ClinicalConfidenceFlags,
    ClinicalConfidenceInput,
    ClinicalConfidenceResult,
    ConfidencePolicy,
    HallucinationRisk,
    ProviderConsistencyResult,
)
from .uncertainty_score import calculate_uncertainty_score


class ClinicalConfidenceEngine:
    def __init__(
        self,
        flags: ClinicalConfidenceFlags | None = None,
        policy: ConfidencePolicy = DEFAULT_CONFIDENCE_POLICY,
        audit_sink: InMemoryConfidenceAuditSink | None = None,
    ) -> None:
        self.flags = flags or load_clinical_confidence_flags({})
        self.policy = policy
        self.audit_sink = audit_sink or InMemoryConfidenceAuditSink()

    def evaluate(self, item: ClinicalConfidenceInput) -> ClinicalConfidenceResult:
        evidence = evaluate_evidence(item)
        provider_consistency = (
            evaluate_provider_consistency(item.provider_outputs)
            if self.flags.provider_consistency_enabled
            else ProviderConsistencyResult(0, 1.0, False, [])
        )
        multimodal = detect_multimodal_conflict(item) if self.flags.multimodal_enabled else None
        hallucination = (
            estimate_hallucination_risk(evidence, provider_consistency, multimodal.multimodal_conflict_detected if multimodal else False)
            if self.flags.hallucination_check_enabled
            else HallucinationRisk("low", False, False, False, False, False)
        )
        uncertainty = calculate_uncertainty_score(evidence, provider_consistency, hallucination)
        confidence = calculate_confidence_score(evidence, provider_consistency, uncertainty, hallucination)
        escalation = build_escalation_recommendation(
            confidence,
            uncertainty,
            hallucination,
            evidence.evidence_completeness,
            provider_consistency.unresolved_conflicts,
            multimodal.conflicts if multimodal else [],
            self.policy,
        )
        safe_display = evaluate_safe_display(
            confidence,
            uncertainty,
            hallucination,
            escalation.escalation_recommended,
            self.flags,
            self.policy,
        )
        result = ClinicalConfidenceResult(
            trace_id=item.trace_id,
            confidence_score=confidence,
            uncertainty_score=uncertainty,
            hallucination_risk=hallucination,
            evidence_completeness=evidence.evidence_completeness,
            provider_consistency=provider_consistency,
            multimodal_conflict_detected=multimodal.multimodal_conflict_detected if multimodal else False,
            escalation_recommended=bool(self.flags.enabled and self.flags.auto_escalation_enabled and escalation.escalation_recommended),
            escalation_reason=escalation.escalation_reason,
            safe_to_display=safe_display.safe_to_display,
            display_restrictions=safe_display.display_restrictions,
            confidence_explanation=[],
            audit_ref=f"{item.trace_id}:clinical-confidence",
            created_at=datetime.now(UTC).isoformat(),
        )
        result.confidence_explanation = explain_confidence(result, evidence)
        self.audit_sink.append(create_confidence_audit_event(result))
        return result

    def audit_events(self):
        return self.audit_sink.list()
